src/main/java/jenkins/branch/WorkspaceLocatorImpl.java

```diff
@@ -444,6 +446,64 @@ private static synchronized void taskFinished() {
             Deleter.class.notifyAll();
         }
 
+        private static class CleanupTaskProvisioner implements Runnable{
+
+            @NonNull
+            private final TopLevelItem tli;
+
+            @NonNull
+            private final Queue<Node> nodes;
+
+            private final int threadLimit = getThreadLimit();
+
+            public CleanupTaskProvisioner(TopLevelItem tli, List<Node> nodes) {
+                this.tli = tli;
+                this.nodes = new LinkedList<>(nodes);
+                if (threadLimit > 0) {
+                    LOGGER.log(Level.INFO, "ThreadLimit Loaded from Environment : {0}", threadLimit);
+                }
+            }
+
+            @Override
+            public void run() {
+                try {
+                    while (!nodes.isEmpty()){
+                        if (hasFreeThreadVolume(threadLimit)) {
+                            Computer.threadPoolForRemoting.submit(new CleanupTask(tli, nodes.remove()));
+                        } else {
+                            Thread.sleep(1000);
+                        }
+                        LOGGER.log(Level.INFO, "LIVE THREAD COUNT : {0}", ManagementFactory.getThreadMXBean().getThreadCount());
+                    }
+                } catch (Exception e) {
+                    LOGGER.log(Level.WARNING, e.getMessage());
+                }
+            }
+
+            private boolean hasFreeThreadVolume(int threadLimit){
+                if (threadLimit <= 0) {
+                    return true;
+                }
+                int currentThreadCount = ManagementFactory.getThreadMXBean().getThreadCount();
+
+                return currentThreadCount < threadLimit;
+            }
+
+            private int getThreadLimit() {
+                String branchApiThreadLimit = System.getenv("BRANCH_API_THREAD_LIMIT");
```

Comment:
Can we put this at the top of the class with the other system properties and do something like:
```suggestion
                static final int = Integer.getInteger(WorkspaceLocatorImpl.class.getName() + "BRANCH_API_THREAD_LIMIT", 0);
```

Reply:
Hey @car-roll, happy to make more adjustments where necessary.
Tidying up the thread limit property is a good idea, but it's used in `CleanupTaskProvisioner`, an internal class of `Deleter`, which is a static class. I'm not sure how to combine that with the suggestion.

Do I move the property to `Deleter` and pass it into `CleanupTaskProvisioner`'s constructor (suggestion below), or is there some other way of achieving what you're after?

```java
    @Extension
    public static class Deleter extends ItemListener {

        private static final String CLEANUP_THREAD_LIMIT = System.getProperty(Deleter.class.getName() + "CLEANUP_THREAD_LIMIT", 0);
        
        @Override
        public void onDeleted(Item item) {
            // <omitted code>

            // Starts provisioner Thread which is tasked with starting cleanup Threads
            new CleanupTaskProvisioner(tli, jenkins.getNodes(), CLEANUP_THREAD_LIMIT).run();
        }
```

Reply:
That looks good 👍

Reply:
Hey @car-roll, I just pushed the changes to replace the environment variable with a system property.

I jammed a small function into `Deleter` to help parse it to an integer. There's a similar function in `org.apache.commons.lang3.NumberUtils`, but I didn't quite want to fiddle with the project's dependencies. Any opinions there?